src/main/java/net/mcreator/ui/modgui/ModElementChangedListener.java

```diff
@@ -0,0 +1,139 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.ui.modgui;
+
+import javafx.embed.swing.JFXPanel;
+import net.mcreator.ui.component.JItemListField;
+import net.mcreator.ui.minecraft.JEntriesList;
+import net.mcreator.ui.minecraft.MCItemHolder;
+import net.mcreator.ui.modgui.codeviewer.ModElementCodeViewer;
+
+import javax.swing.*;
+import javax.swing.event.ChangeEvent;
+import javax.swing.event.ChangeListener;
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import javax.swing.text.JTextComponent;
+import java.awt.*;
+import java.awt.event.*;
+
+/**
+ * Gets triggered whenever a change to a mod element is detected inside {@link ModElementGUI}
+ * @see ModElementGUI#elementUpdateListener
+ * @see ModElementCodeViewer#codeChangedListener
+ */
+public interface ModElementChangedListener
+		extends MouseListener, KeyListener, ActionListener, ChangeListener, DocumentListener {
+
+	/**
+	 * The main listener method, triggered when an event occurs on a registered component
+	 */
+	void modElementChanged();
+
+	/**
+	 * Registers the given UI component to trigger this listener when a change is detected on it
+	 *
+	 * @param container The UI element to register
+	 * @param addIdleListeners Specifies whether to track mouse/keyboard events for this container
+	 */
+	default void registerUI(JComponent container, boolean addIdleListeners) {
+		for (Component component : container.getComponents()) {
+			if (component instanceof MCItemHolder itemHolder) {
+				itemHolder.addBlockSelectedListener(this);
+			} else if (component instanceof JItemListField<?> listField) {
+				listField.addChangeListener(this);
+			} else if (component instanceof JEntriesList entriesList) {
+				registerUI(entriesList, addIdleListeners);
+				entriesList.addEntryRegisterListener(c -> {
+					registerUI(c, addIdleListeners);
+					modElementChanged();
+				});
+				if (addIdleListeners)
+					component.addMouseListener(this);
+			} else if (component instanceof AbstractButton button) {
+				button.addActionListener(this);
+			} else if (component instanceof JSpinner spinner) {
+				spinner.addChangeListener(this);
+			} else if (component instanceof JComboBox<?> comboBox) {
+				comboBox.addActionListener(this);
+			} else if (component instanceof JTextComponent textComponent) {
+				textComponent.getDocument().addDocumentListener(this);
+			} else if (component instanceof JFXPanel) {
+				component.addMouseListener(this);
+				component.addKeyListener(this);
+			} else if (component instanceof JComponent jcomponent) {
+				registerUI(jcomponent, addIdleListeners);
+
+				if (!(component instanceof JLabel) && !(component instanceof JPanel) && addIdleListeners) {

```

Comment:
This could exclude elements that change GE when addIdleListeners is false. GE is changed when code is changed, so I don't see how listening for GE changes would need addIdleListeners to false and for code changes addIdleListeners to true

Reply:
The way you designed the preview generator, it can sometimes ignore GE change in cases of "overflow" (`reload()` being called too fast); then, idle events (mouse clicks or key pressing over empty panels) would attempt once again to update element code.
But, speaking of GE change on idle events... Well, what if you open a `CodeEditorView`, just change the caret position and this marks file as changed?

Reply:
Mod elements don't really have CodeEditorView in them, except for function.

 But say some custom element changes on the key event, setting this to false will make this change ignored.

However we take it, code preview and GE changed should trigger on same events, no reason for a flag here

Reply:
>Mod elements don't really have CodeEditorView in them

I spoke abstract...